### PR TITLE
[Linux] Add new test case pvrdma_network_device_ops

### PR DESCRIPTION
**Pull request by keirazhang**
Add a new Linux test case pvrdma_network_device_ops to test pvrdma network adapter.

Test results on vSphere 7.0
```
- - 'VCENTER: 7.0.0 ob-15952498'
  - 'ESXI: 7.0.0 ob-15843807'
  - 'GUEST OS: RedHat 9.1 x86_64'
  - 'Results: Total (4), Passed (4)'
  - ' * deploy_vm_efi_paravirtual_vmxnet3...............................Passed'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops.......................................Passed'
  - 
- - 'VCENTER: 7.0.0 ob-15952498'
  - 'ESXI: 7.0.0 ob-15843807'
  - 'GUEST OS: Ubuntu 22.10 x86_64'
  - 'Results: Total (4), Passed (4)'
  - ' * deploy_vm_efi_paravirtual_vmxnet3...............................Passed'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops.......................................Passed'

- - 'VCENTER: 7.0.0 ob-15952498'
  - 'ESXI: 7.0.0 ob-15843807'
  - 'GUEST OS: CentOS 8.5 x86_64'
  - 'Results: Total (4), Passed (4)'
  - ' * deploy_vm_efi_paravirtual_vmxnet3...............................Passed'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops.......................................Passed'


- - 'VCENTER: 7.0.0 ob-15952498'
  - 'ESXI: 7.0.0 ob-15843807'
  - 'GUEST OS: Debian 11.6 i386'
  - 'Results: Total (4), Passed (3), Skipped (1)'
  - ' * deploy_vm_efi_paravirtual_vmxnet3...............................Passed'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops................................Not Supported'

- - 'VCENTER: 7.0.0 ob-15952498'
  - 'ESXI: 7.0.0 ob-15843807'
  - 'GUEST OS: Ubuntu 20.04 x86_64'
  - 'Results: Total (4), Passed (4)'
  - ' * deploy_vm_efi_paravirtual_vmxnet3...............................Passed'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops.......................................Passed'

- - 'VCENTER: 7.0.0 ob-15952498'
  - 'ESXI: 7.0.0 ob-15843807'
  - 'GUEST OS: SLES 15.5 x86_64'
  - 'Results: Total (4), Passed (4)'
  - ' * deploy_vm_efi_paravirtual_vmxnet3...............................Passed'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops.......................................Passed'

  - 'GUEST OS: Rocky 8.7 x86_64'
  - 'Hardware Version: vmx-17'
  - 'GuestInfo Guest Id: other4xLinux64Guest'
  - 'Results: Total (4), Passed (4)'
  - ' * deploy_vm_efi_paravirtual_vmxnet3...............................Passed'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops.......................................Passed'

- - 'VCENTER: 7.0.0 ob-15952498'
  - 'ESXI: 7.0.0 ob-15843807'
  - 'GUEST OS: OracleLinux 9.1 x86_64'
  - 'Results: Total (4), Passed (4)'
  - ' * deploy_vm_efi_paravirtual_vmxnet3...............................Passed'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops.......................................Passed'

- - 'VCENTER: 7.0.0 ob-15952498'
  - 'ESXI: 7.0.0 ob-15843807'
  - 'GUEST OS: VMware Photon OS 4.0 x86_64'
  - 'Results: Total (4), Passed (3), Skipped (1)'
  - ' * deploy_vm_efi_paravirtual_vmxnet3...............................Passed'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops................................Not Supported'
```

### CODE REVIEW

File: common/esxi_set_firewall_rule.yml [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    that:
+      - rule_name is defined and rule_name
+      - rule_enabled is defined
+      - rule_enabled | type_debug == 'bool'
+    fail_msg: "At least one of parameters 'rule_name' or 'rule_enabled' is incorrect"
+
+- name: "Check parameter for enabling ESXi host firewall rule '{{ rule_name }}'"
+  ansible.builtin.assert:
+    that:
+      - rule_allowed_all is defined
```
Comment (Tomorrow9) [inline common/esxi_set_firewall_rule.yml:23]: I think here `rule_allowed_all` can be set according to the parameters `rule_allowed_ip` or `rule_allowed_networks`, if they're set then it's false, if they're not set then it's true. Also these 2 optional paramter value can also be checked.

File: common/vcenter_manage_vds_host.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    state: "{{ vds_host_state }}"
+    vmnics: "{{ vds_host_vmnics | default(omit) }}"
+  register: manage_vds_host_result
+
+- name: "Check the result of managing ESXi host for vSphere Distributed Switch"
+  ansible.builtin.assert:
+    that:
+      - manage_vds_host_result is defined
+      - manage_vds_host_result.failed is defined
+      - not manage_vds_host_result.failed
```
Comment (Tomorrow9) [inline common/vcenter_manage_vds_host.yml:49]: So does here.

File: common/vcenter_manage_vds_portgroup.yml [hunk 3]
```diff
@@ -0,0 +1,83 @@
+# Copyright 2023 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+# Create or remove a vSphere Distributed Port Group
+# Parameters:
+#   vds_name: The vSphere Distributed Switch name where the port group is.
+#   vds_portgroup_name: The distributed port group name.
+#   vds_portgroup_state: The distributed port group state: present or absent.
+#   vds_port_number: The number of ports the distributed port group should contain.
```
Comment (Tomorrow9) [inline common/vcenter_manage_vds_portgroup.yml:9]: This parameter can also have a default value '4'

File: common/esxi_configure_advanced_settings.yml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    esxi_hostname: "{{ esxi_hostname }}"
+    options: "{{ esxi_advanced_settings }}"
+  register: esxi_advanced_settings_result
+
+- name: "Check advanced settings are configured"
+  ansible.builtin.assert:
+    that:
+      - esxi_advanced_settings_result is defined
+      - esxi_advanced_settings_result.failed is defined
+      - not esxi_advanced_settings_result.failed
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply common/esxi_configure_advanced_settings.yml:30]: If the advanced settings had been set already, changed would be false, but the task is still successful. So here checks fails.

File: common/esxi_manage_vmkernel.yml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    device: "{{ vmk_device | default(omit) }}"
+    network: "{{ vmk_network | default(omit) }}"
+    enable_mgmt: "{{ vmk_enable_mgmt | default(false) }}"
+  register: manage_vmk_result
+
+- name: "Check the result of managing VMkernel network adapter"
+  ansible.builtin.assert:
+    that:
+      - manage_vmk_result is defined
+      - manage_vmk_result.failed is defined
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply common/esxi_manage_vmkernel.yml:58]: Same reason as above.

File: common/README.md [hunk 6]
```diff
@@ -120,6 +121,11 @@
 * esxi_get_guest_ids.yml: Get supported guest ids for a hardware version on ESXi host
 * esxi_get_guest_config_options.yml: Get VM config options for a guest id on a hardware version
 * esxi_get_pmem_info.yml: Get ESXi host total and available persistent memroy size in MB
+* esxi_configure_advanced_settings.yml: Configure advanced system settings on ESXi host
+* esxi_get_vmkernels.yml: Get ESXi host VMkernel network adapters
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply common/README.md]: Files are renamed. Thanks~

File: common/esxi_set_firewall_rule.yml (hunk 1 above)
Comment (keirazhang) [reply common/esxi_set_firewall_rule.yml]: Updated.

File: common/vcenter_manage_vds_portgroup.yml (hunk 3 above)
Comment (keirazhang) [reply common/vcenter_manage_vds_portgroup.yml:9]: Updated.

File: common/esxi_configure_advanced_settings.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (keirazhang) [reply common/esxi_configure_advanced_settings.yml]: Added separate tasks to check the result based on context. Here is changed to only display the task result.

File: common/esxi_manage_vmkernel.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (keirazhang) [reply common/esxi_manage_vmkernel.yml]: Added separate tasks to check the result based on context. Here is changed to only display the task result.

File: common/esxi_manage_vmkernel_adapter.yml [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+# port group or a vSphere Distributed Switch (VDS) port group
+# Paramters:
+#   vmk_device: The VMkernel network adapter name.
+#   vmk_state: The VMkernel network adapter state: present or absent.
+#   vmk_vss_name: The vSphere Standard Switch name for the VMkernel network adapter.
+#   vmk_vds_name: The vSphere Distributed Switch name for the VMkernel network adapter.
+#   vmk_portgroup_name: The port group name for the VMkernel network adapter.
+#   vmk_network: The network config for the VMkernel network adapter.
+#   vmk_enable_mgmt: True to enable management on the VMkernel network adapter.
+#     Default value if false.
```
Comment (123lzxm) [inline common/esxi_manage_vmkernel_adapter.yml:14]: "Default value if false." ---> "Default value is false."

File: common/vcenter_manage_vds_host.yml (hunk 2 above)
Comment (keirazhang) [reply common/vcenter_manage_vds_host.yml]: Updated the result checking.

File: linux/network_device_ops/check_pvrdma_support_status.yml [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  include_tasks: ../../common/skip_test_case.yml
+  vars:
+    skip_msg: >-
+      Test case '{{ ansible_play_name }}' is not supported because {{ guest_os_ansible_distribution }}
+      {{ guest_os_ansible_distribution_ver }} doesn't have vmw_pvrdma module.
+    skip_reason: "Not Supported"
+  when: >
+    (guest_pvrdma_module_info.filename is undefined) or
+    (('vmw_pvrdma' not in guest_pvrdma_module_info.filename) and
+    ('builtin' not in guest_pvrdma_module_info.filename))
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply linux/network_device_ops/check_pvrdma_support_status.yml]: Photon OS has drivers built in its kernel, so its drivers filenames are builtin in Photon OS 4.x. But it doesn't support pvrdma. So this line can be removed.

File: linux/network_device_ops/pvrdma_network_device_ops.yml [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+  tasks:
+    - name: "Test case block"
+      block:
+        - include_tasks: ../setup/test_setup.yml
+
+        - include_tasks: check_pvrdma_support_status.yml
+
+        - include_tasks: prepare_network_device_test.yml
+
+        - include_tasks: ../utils/get_rdma_device_info.yml
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply linux/network_device_ops/pvrdma_network_device_ops.yml]: No necessary to add name for test_setup. The others include_tasks have been added name.

File: linux/utils/reload_module.yml [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  ansible.builtin.assert:
+    that:
+      - module_name is defined and module_name
+    fail_msg: "module_name variable must be defined and set to a valid value."
+
+- name: "Check module '{{ module_name }}' is loaded or not"
+  ansible.builtin.shell: "lsmod | grep '^{{ module_name }}'"
+  delegate_to: "{{ vm_guest_ip }}"
+  ignore_errors: true
+  changed_when: false
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply linux/utils/reload_module.yml:18]: No need to check changed or not here. Here only need the output from the command.

File: common/esxi_manage_vmkernel_adapter.yml (hunk 7 above)
Comment (keirazhang) [reply common/esxi_manage_vmkernel_adapter.yml]: Fixed

File: common/vds_network_setup.yml [hunk 11]
```diff
@@ -0,0 +1,76 @@
+# Copyright 2023 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+# Setup vSphere Distributed Switch (VDS) networking, include:
+#   - Create a new vSphere Distributed Switch
+#   - Create a new distributed port group
+#   - Add ESXi host to the distributed port group
```
Comment (Tomorrow9) [inline common/vds_network_setup.yml:7]: Add ESXi host to the distributed switch

File: common/README.md [hunk 12]
```diff
@@ -137,15 +146,21 @@
...
-* esxi_remove_vswitch.yml: Remove a standard virtual switch
+* esxi_add_vswitch.yml: Add a new vSphere Standard Switch
+* esxi_remove_vswitch.yml: Remove a vSphere Standard Switch
 * esxi_get_vswitch_facts.yml: Get virtual switch facts
 
+### Tasks for vSphere Distributed Switch (VDS) and Port Group
+* vcenter_manage_vds.yml: Create or remove a vSphere Distributed Switch
+* vcenter_get_vds_info.yml: Gather information about a vSphere Distributed Switch
+* vcenter_manage_vds_portgroup.yml: Create or remove a distributed port group on vCenter Server
+* vcenter_manage_vds_host.yml: Add or remove an ESXi host for a distributed port group on vCenter Server
```
Comment (Tomorrow9) [inline common/README.md:162]: to distributed switch

File: common/vds_network_setup.yml [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+  include_tasks: vcenter_manage_vds.yml
+  vars:
+    vds_state: "present"
+
+- name: "Create a new distributed port group"
+  include_tasks: vcenter_manage_vds_portgroup.yml
+  vars:
+    vds_portgroup_state: "present"
+
+- name: "Add ESXi host to the distributed port group"
```
Comment (Tomorrow9) [inline common/vds_network_setup.yml:24]: distributed switch

File: common/esxi_enable_pvrdma_configs.yml [hunk 14]
```diff
@@ -0,0 +1,30 @@
+# Copyright 2023 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+# Configure the VMkernel adapter and firewall rule on ESXi host for PVRDMA communication.
+# Parameters:
+#   pvrdma_vmk_nic: The VMkernel network adapter to tag for PVRDMA communication
+#
+
+- name: "Tag VMkernel adapter '{{ pvrdma_vmk_nic }}' for PVRDMA"
+  include_tasks: ../../common/esxi_set_advanced_system_settings.yml
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply common/esxi_enable_pvrdma_configs.yml:10]: They are completely different tasks. `esxi_set_advanced_system_settings.yml` is to configure ESXi server's advanced system settings. And `vm_set_extra_config.yml` is to configure VM's advanced settings.

File: common/esxi_enable_pvrdma_configs.yml [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+  vars:
+    esxi_host_config_options: {"Net.PVRDMAVmknic": "{{ pvrdma_vmk_nic }}"}
+
+- name: "Get ESXi server's advanced system settings to check config result"
+  include_tasks: esxi_get_advanced_system_settings.yml
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply common/esxi_enable_pvrdma_configs.yml:15]: One is for ESXi host, and the other is for VM.

File: common/esxi_enable_pvrdma_configs.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (123lzxm) [reply common/esxi_enable_pvrdma_configs.yml:10]: Ok. Got it.

File: common/README.md (hunk 12 above)
Comment (keirazhang) [reply common/README.md]: updated.

File: linux/network_device_ops/pvrdma_network_device_ops.yml [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+
+        # On server VM
+        - name: "Initialize the static IP address for the server VM"
+          ansible.builtin.set_fact:
+            new_nic_network: "{{ (pvrdma_server_vm_ipv4 ~ '/' ~ pvrdma_server_vm_net_prefix) | ansible.utils.ipaddr('network') }}"
+            new_nic_ipv4: "{{ pvrdma_server_vm_ipv4 }}"
+            new_nic_net_prefix: "{{ pvrdma_server_vm_net_prefix }}"
+            new_nic_gateway: "{{ pvrdma_server_vm_gateway }}"
+
+        - name: "Hot add a new network adapter on server VM and apply network config"
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply linux/network_device_ops/pvrdma_network_device_ops.yml]: updated

File: linux/network_device_ops/pvrdma_network_device_ops.yml [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+          ansible.builtin.assert:
+            that:
+              - rping_server_result is defined
+              - rping_server_result.rc is defined
+              - rping_server_result.rc == 0
+            fail_msg: >-
+              RDMA pinging server VM from client VM failed.
+              Hit error '{{ rping_server_result.stderr | default() }}'
+
+        - name: "Power off the cloned VM"
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply linux/network_device_ops/pvrdma_network_device_ops.yml]: updated

File: common/vds_network_setup.yml (hunk 11 above)
Comment (keirazhang) [reply common/vds_network_setup.yml]: updated

File: common/esxi_get_vmkernel_adapters.yml [hunk 18]
```diff
@@ -0,0 +1,31 @@
+# Copyright 2023 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+# Gather info about ESXi VMkernel adapters
+# Return:
+#   esxi_vmkernels_info: A list of VMkernel adapters info
+
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply common/esxi_get_vmkernel_adapters.yml]: updated.

**Review by Tomorrow9 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**